File: web/admin/src/api/inbox.ts
import axios from 'axios';
import { APIResponse } from '@/api/types';

export interface Inbox {
  id: string;
  title?: string;
  description?: string;
  max_items: number;
  is_public: boolean;
}

export interface SystemAuthToken {
  id: number;
  token: string;
  label?: string;
  created_at: string;
}

const adminApiBase = '/api/admin';

// Inboxes CRUD
export function createInbox(data: Inbox): Promise<APIResponse<Inbox>> {
  return axios
    .post<APIResponse<Inbox>>(`${adminApiBase}/inboxes`, data)
    .then((res) => res.data);
}

export function listInboxes(): Promise<APIResponse<Inbox[]>> {
  return axios
    .get<APIResponse<Inbox[]>>(`${adminApiBase}/inboxes`)
    .then((res) => res.data);
}

export function getInbox(id: string): Promise<APIResponse<Inbox>> {
  return axios
    .get<APIResponse<Inbox>>(`${adminApiBase}/inboxes/${id}`)
    .then((res) => res.data);
}

export function updateInbox(
  id: string,
  data: Inbox
): Promise<APIResponse<Inbox>> {
  return axios
    .put<APIResponse<Inbox>>(`${adminApiBase}/inboxes/${id}`, data)
    .then((res) => res.data);
}

export function deleteInbox(id: string): Promise<APIResponse<void>> {
  return axios
    .delete<APIResponse<void>>(`${adminApiBase}/inboxes/${id}`)
    .then((res) => res.data);
}

// System Auth Tokens CRUD
export function createSystemAuthToken(data: {
  label: string;
}): Promise<APIResponse<SystemAuthToken>> {
  return axios
    .post<APIResponse<SystemAuthToken>>(
      `${adminApiBase}/system-auth-tokens`,
      data
    )
    .then((res) => res.data);
}

export function listSystemAuthTokens(): Promise<
  APIResponse<SystemAuthToken[]>
> {
  return axios
    .get<APIResponse<SystemAuthToken[]>>(`${adminApiBase}/system-auth-tokens`)
    .then((res) => res.data);
}

export function deleteSystemAuthToken(id: number): Promise<APIResponse<void>> {
  return axios
    .delete<APIResponse<void>>(`${adminApiBase}/system-auth-tokens/${id}`)
    .then((res) => res.data);
}

File: internal/controller/inbox.go
package controller

import (
	"errors"
	"net/http"

	"FeedCraft/internal/dao"
	"FeedCraft/internal/util"

	"github.com/gin-gonic/gin"
	"github.com/google/uuid"
	"gorm.io/gorm"
)

func CreateInbox(c *gin.Context) {
	var data dao.Inbox
	if err := c.ShouldBindJSON(&data); err != nil {
		c.JSON(http.StatusBadRequest, util.APIResponse[any]{Msg: err.Error()})
		return
	}
	db := util.GetDatabase()

	if err := dao.CreateInbox(db, &data); err != nil {
		c.JSON(http.StatusInternalServerError, util.APIResponse[any]{Msg: err.Error()})
		return
	}
	c.JSON(http.StatusCreated, util.APIResponse[any]{Data: data})
}

func GetInbox(c *gin.Context) {
	id := c.Param("id")
	db := util.GetDatabase()

	data, err := dao.GetInboxByID(db, id)
	if err != nil {
		if errors.Is(err, gorm.ErrRecordNotFound) {
			c.JSON(http.StatusNotFound, util.APIResponse[any]{Msg: "Inbox not found"})
			return
		}
		c.JSON(http.StatusInternalServerError, util.APIResponse[any]{Msg: err.Error()})
		return
	}
	c.JSON(http.StatusOK, util.APIResponse[any]{Data: data})
}

func UpdateInbox(c *gin.Context) {
	id := c.Param("id")
	var data dao.Inbox
	if err := c.ShouldBindJSON(&data); err != nil {
		c.JSON(http.StatusBadRequest, util.APIResponse[any]{Msg: err.Error()})
		return
	}
	data.ID = id
	db := util.GetDatabase()

	if _, err := dao.GetInboxByID(db, id); err != nil {
		if errors.Is(err, gorm.ErrRecordNotFound) {
			c.JSON(http.StatusNotFound, util.APIResponse[any]{Msg: "Inbox not found"})
			return
		}
		c.JSON(http.StatusInternalServerError, util.APIResponse[any]{Msg: err.Error()})
		return
	}

	if err := dao.UpdateInbox(db, &data); err != nil {
		c.JSON(http.StatusInternalServerError, util.APIResponse[any]{Msg: err.Error()})
		return
	}
	c.JSON(http.StatusOK, util.APIResponse[any]{Data: data})
}

func DeleteInbox(c *gin.Context) {
	id := c.Param("id")
	db := util.GetDatabase()

	if err := dao.DeleteInbox(db, id); err != nil {
		if errors.Is(err, gorm.ErrRecordNotFound) {
			c.JSON(http.StatusNotFound, util.APIResponse[any]{Msg: "Inbox not found"})
			return
		}
		c.JSON(http.StatusInternalServerError, util.APIResponse[any]{Msg: err.Error()})
		return
	}
	c.JSON(http.StatusOK, util.APIResponse[any]{})
}

func ListInboxes(c *gin.Context) {
	db := util.GetDatabase()
	list, err := dao.ListInboxes(db)
	if err != nil {
		c.JSON(http.StatusInternalServerError, util.APIResponse[any]{Msg: err.Error()})
		return
	}
	c.JSON(http.StatusOK, util.APIResponse[any]{Data: list})
}

// SystemAuthToken Admin CRUD

func CreateSystemAuthToken(c *gin.Context) {
	var data struct {
		Label string `json:"label"`
	}
	if err := c.ShouldBindJSON(&data); err != nil {
		c.JSON(http.StatusBadRequest, util.APIResponse[any]{Msg: err.Error()})
		return
	}

	tokenValue := uuid.New().String()
	tokenRecord := dao.SystemAuthToken{
		Token: tokenValue,
		Label: data.Label,
	}

	db := util.GetDatabase()
	if err := dao.CreateSystemAuthToken(db, &tokenRecord); err != nil {
		c.JSON(http.StatusInternalServerError, util.APIResponse[any]{Msg: err.Error()})
		return
	}
	c.JSON(http.StatusCreated, util.APIResponse[any]{Data: tokenRecord})
}

func ListSystemAuthTokens(c *gin.Context) {
	db := util.GetDatabase()
	list, err := dao.ListSystemAuthTokens(db)
	if err != nil {
		c.JSON(http.StatusInternalServerError, util.APIResponse[any]{Msg: err.Error()})
		return
	}
	c.JSON(http.StatusOK, util.APIResponse[any]{Data: list})
}

func DeleteSystemAuthToken(c *gin.Context) {
	idParam := c.Param("id")
	db := util.GetDatabase()

	if err := db.Where("id = ?", idParam).Delete(&dao.SystemAuthToken{}).Error; err != nil {
		c.JSON(http.StatusInternalServerError, util.APIResponse[any]{Msg: err.Error()})
		return
	}
	c.JSON(http.StatusOK, util.APIResponse[any]{})
}
